packages/teleport/src/DesktopSession/TdpClientCanvas.tsx

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import React, { useEffect, useRef, CSSProperties } from 'react';
+import { RenderData } from 'teleport/lib/tdp/client';
+import useTdpClientCanvas from './useTdpClientCanvas';
+
+export default function TdpClientCanvas(props: Props) {
+  const {
+    tdpClient,
+    onInit,
+    onConnect,
+    onRender,
+    onDisconnect,
+    onError,
+    onKeyDown,
+    onKeyUp,
+    onMouseMove,
+    onMouseDown,
+    onMouseUp,
+    onResize,
+    style,
+  } = props;
+
+  const canvasRef = useRef<HTMLCanvasElement>(null);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const ctx = canvas.getContext('2d');
+
+    // Buffered rendering logic
+    var buffer: RenderData[] = [];
+    const renderBuffer = () => {
+      if (buffer.length) {
+        for (let i = 0; i < buffer.length; i++) {
+          onRender(ctx, buffer[i]);
+        }
+        buffer = [];
+      }
+      requestAnimationFrame(renderBuffer);
+    };
+    requestAnimationFrame(renderBuffer);
```

Comment:
i am not sure I understand if this buffering is ever used (accept on very first run)

Reply:
The other straightforward option is to put it on a timer and then `requestAnimationFrame` every N ms, in this case we're just drawing at the highest rate possible when the buffer is being filled.

I could build a system that stops requesting frames if the buffer has been idle for N seconds, but this seems to work for now. Is there an easy way to measure cpu usage of those calls?

Reply:
This is a standard use of the api, it is shown in [the MDN example](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame#example) and is [common for game loops](https://spicyyoghurt.com/tutorials/html5-javascript-game-development/create-a-proper-game-loop-with-requestanimationframe). 

It was originally built to replace `setInterval` loops and includes [browser optimizations](https://blog.teamtreehouse.com/efficient-animations-with-requestanimationframe). What I have here is basically like a recursive `setTimeout` with the timeout matched to the display refresh rate rather than hardcoded.